### PR TITLE
Add shipping box type

### PR DESCRIPTION
**Pull request by usudaysingh**

### CODE REVIEW

**Comment by usudaysingh** [comment]
screen shot 2015-12-18 at 1 31 46 am

File: shipment.xml [hunk 1]
```diff
@@ -65,5 +65,40 @@
...
+            <field name="type">tree</field>
+            <field name="name">box_type_tree</field>
+        </record>
+
+        <record model="ir.action.act_window" id="act_box_type_form">
+            <field name="name">Shipping Box Type</field>
+            <field name="res_model">shipment.box_types</field>
+        </record>
+
+	    <record model="ir.action.act_window.view" id="act_uom_form_view1">
```
Comment (priyankarani) [inline shipment.xml]: Uom_form?

File: shipment.xml [hunk 2]
```diff
@@ -65,5 +65,40 @@
...
+        </record>
+
+        <record model="ir.action.act_window.view" id="act_uom_form_view2">
+            <field name="sequence" eval="20"/>
+            <field name="view" ref="box_type_form"/>
+            <field name="act_window" ref="act_box_type_form"/>
+        </record>
+        
+        <menuitem name="Settings" id="menu_settings" icon="tryton-preferences-system"/>
+        <menuitem name="Inventory &amp; Stock" id="menu_stock" parent="menu_settings" sequence="100"/>
```
Comment (priyankarani) [inline shipment.xml]: Settings is inside fulfilio. You need to implement this menu there. Here move it inside inventory and stock

File: shipment.py [hunk 3]
```diff
@@ -511,3 +516,19 @@ def generate_label(self, shipment):
...
+    __name__ = 'shipment.box_types'
+
+    carrier = fields.Selection(CARRIERS, 'Carrier', required=True)
+    code = fields.Char('Code', required=True)
+    length = fields.Float('Length')
+    width = fields.Float('Width')
+    height = fields.Float('Height')
+    distance_unit = fields.Many2One(
+        'product.uom', 'Distance Unit', states={
+           'required': Or(Bool(Eval('length')), Bool(Eval('width')), Bool(Eval('height')))
```
Comment (priyankarani) [inline shipment.py]: Add domain for category

File: shipment.xml [hunk 4]
```diff
@@ -65,5 +65,39 @@
...
+            <field name="act_window" ref="act_box_type_form"/>
+        </record>
+
+        <record model="ir.action.act_window.view" id="act_box_type_view2">
+            <field name="sequence" eval="20"/>
+            <field name="view" ref="box_type_form"/>
+            <field name="act_window" ref="act_box_type_form"/>
+        </record>
+        
+        <menuitem name="Inventory &amp; Stock" id="menu_stock" parent="menu_settings" sequence="100"/>
```
Comment (priyankarani) [inline shipment.xml]: not this. check for inventory stock menu in stock module

File: shipment.py [hunk 5]
```diff
@@ -511,3 +516,20 @@ def generate_label(self, shipment):
...
+
+    carrier = fields.Selection(CARRIERS, 'Carrier', required=True)
+    code = fields.Char('Code', required=True)
+    length = fields.Float('Length')
+    width = fields.Float('Width')
+    height = fields.Float('Height')
+    distance_unit = fields.Many2One(
+        'product.uom', 'Distance Unit', states={
+            'required': Or(Bool(Eval('length')), Bool(
+                Eval('width')), Bool(Eval('height')))
```
Comment (priyankarani) [inline shipment.py]: add domain https://github.com/fulfilio/product-variant-measurements/blob/develop/product.py#L28. So it only pick uom related to length

**Comment by usudaysingh** [comment]
screen shot 2015-12-18 at 2 18 24 am

File: shipment.py [hunk 6]
```diff
@@ -511,3 +516,23 @@ def generate_label(self, shipment):
...
+            },
+        domain=[
+            ('category', '=', Id('product', 'uom_cat_length'))
+        ],
+    )
```
Comment (priyankarani) [inline shipment.py]: Depend missing

File: shipment.py [hunk 7]
```diff
@@ -511,3 +516,23 @@ def generate_label(self, shipment):
...
+    width = fields.Float('Width')
+    height = fields.Float('Height')
+    distance_unit = fields.Many2One(
+        'product.uom', 'Distance Unit', states={
+            'required': Or(Bool(Eval('length')), Bool(
+                Eval('width')), Bool(Eval('height')))
+            },
+        domain=[
+            ('category', '=', Id('product', 'uom_cat_length'))
+        ], depends=['state']
```
Comment (priyankarani) [inline shipment.py]: what ?? where is state being used ?

File: shipment.py [hunk 8]
```diff
@@ -3,22 +3,27 @@
...
 ]
 
 STATES = {
     'readonly': Eval('state') == 'done',
 }
+CARRIERS = [
+    ('fedex', 'FedEx'),
+    ('ups', 'UPS'),
+    ('usps', 'USPS'),
+    ]
```
Comment (priyankarani) [inline shipment.py]: deindent

File: shipment.py [hunk 9]
```diff
@@ -511,3 +516,23 @@ def generate_label(self, shipment):
             )
 
         return getattr(shipment, method_name)()
+
+
+class ShipmentBoxTypes(ModelSQL, ModelView):
+    "Parcel Box Type"
+    __name__ = 'shipment.box_types'
+
+    carrier = fields.Selection(CARRIERS, 'Carrier', required=True)
```
Comment (tarunbhardwaj) [inline shipment.py]: Selection options should be added by downstream modules.

File: view/box_type_form.xml [hunk 10]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0"?>
+<form string="Shipping Box Type">
+    <label name="code"/>
+    <field name="code"/>
+    <label name="carrier"/>
+    <field name="carrier"/>
+    <label name="length"/>
```
Comment (tarunbhardwaj) [inline view/box_type_form.xml]: better add separator in netween

File: view/box_type_tree.xml [hunk 11]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0"?>
+<tree string="Shipping Box Type">
+    <field name="code"/>
```
Comment (tarunbhardwaj) [inline view/box_type_tree.xml]: add carrier

File: view/box_type_tree.xml [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+    <field name="length"/>
```
Comment (tarunbhardwaj) [inline view/box_type_tree.xml]: no need of length, height, width.

File: shipment.py [hunk 13]
```diff
@@ -3,22 +3,23 @@
...
     'ShipmentOut', 'StockMove', 'GenerateShippingLabelMessage',
     'GenerateShippingLabel', 'ShippingCarrierSelector',
-    'ShippingLabelNoModules', 'Package'
+    'ShippingLabelNoModules', 'Package', 'ShipmentBoxTypes'
 ]
 
 STATES = {
     'readonly': Eval('state') == 'done',
 }
+CARRIERS = []
```
Comment (tarunbhardwaj) [inline shipment.py]: Its not that expensive to create an empty list so please use `[]` directly.

File: view/box_type_form.xml [hunk 14]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0"?>
+<form string="Shipping Box Type" colspan="4">
+    <label name="code"/>
+    <field name="code"/>
+    <label name="carrier"/>
+    <field name="carrier"/>
+    <separator string="Dimensions"  colspan="4"/>
```
Comment (priyankarani) [inline view/box_type_form.xml]: double space, it should have id too

File: shipment.py [hunk 15]
```diff
@@ -3,22 +3,23 @@
...
```
Comment (priyankarani) [inline shipment.py]: fix this

File: shipment.xml [hunk 16]
```diff
@@ -65,5 +65,38 @@
...
+            <field name="type">tree</field>
+            <field name="name">box_type_tree</field>
+        </record>
+
+        <record model="ir.action.act_window" id="act_box_type_form">
+            <field name="name">Shipping Box Type</field>
+            <field name="res_model">shipment.box_types</field>
+        </record>
+
+	    <record model="ir.action.act_window.view" id="act_box_type_view1">
```
Comment (tarunbhardwaj) [inline shipment.xml]: indentation

**Comment by priyankarani** [comment]
LGTM

**Pull request merged**